### PR TITLE
Support type coercion for structural types in Iceberg and Delta Lake

### LINKED ISSUE #21511: Automatic type coercion doesn't work for structural types
Automatic type coercion introduced in Trino 426 (see https://github.com/trinodb/trino/pull/13994) doesn't work for structural types.
We are on Trino 444 now.

Here are tests how to reproduce it, we run them in **iceberg** plugin:
```java
@Test
public void testTypeCoercionMap_create() {
    assertQuerySucceeds("CREATE TABLE type_coercion_map (mp MAP(TIMESTAMP(3), INT))");
}

@Test
public void testTypeCoercionMap_CTAS() {
    assertQuerySucceeds("CREATE TABLE type_coercion_map AS SELECT MAP(ARRAY[localtimestamp], ARRAY[0]) AS mp");
}

@Test
public void testTypeCoercionRow_create() {
    assertQuerySucceeds("CREATE TABLE type_coercion_row (rw ROW(ts TIMESTAMP(3), nb INT))");
}

@Test
public void testTypeCoercionRow_CTAS() {
    assertQuerySucceeds("CREATE TABLE type_coercion_row AS SELECT CAST(ROW(localtimestamp, 0) AS ROW(ts TIMESTAMP(3), nb INT)) AS rw");
}

@Test
public void testTypeCoercionArray_create() {
    assertQuerySucceeds("CREATE TABLE type_coercion_array (arr ARRAY(TIMESTAMP(3)))");
}

@Test
public void testTypeCoercionArray_CTAS() {
    assertQuerySucceeds("CREATE TABLE type_coercion_array AS SELECT ARRAY[localtimestamp] AS arr");
}
```

All above tests fail with:
```
Timestamp precision (3) not supported for Iceberg. Use "timestamp(6)" instead.
```

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #21511
Supersedes #21055 with connector side approach

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake, Iceberg
* Automatically use `varchar` in struct types as a type during table creation when `char` is specified. ({issue}`21511`)

# Iceberg
* Automatically use microsecond precision for temporal types in struct types as a type during table creation. ({issue}`21511`)
```

### CODE REVIEW

**Review by SemionPar — APPROVED** [review]
LGTM!

**Comment by SemionPar** [comment]
To avoid conflicts, basing char coercion in Delta (https://github.com/ebyhr/trino/pull/1) on this branch

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 1]
```diff
@@ -7902,10 +7903,13 @@ private List<TypeCoercionTestSetup> typeCoercionOnCreateTableAsSelectData()
                 .add(new TypeCoercionTestSetup("CHAR 'A '", "varchar", "'A '"))
                 .add(new TypeCoercionTestSetup("CHAR ' A'", "varchar", "' A'"))
                 .add(new TypeCoercionTestSetup("CHAR 'ABc'", "varchar", "'ABc'"))
+                .add(new TypeCoercionTestSetup("ARRAY[CHAR 'A']", "array(varchar)", "ARRAY['A']"))
```
Comment (Praveen2112) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7906]: Can we add test for nested structural types ?

**Review by Praveen2112 — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java (hunk 1 above)
Comment (Praveen2112) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7906]: Is it possible to add them for delta lake ?

Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7906]: The current helper class of TestDeltaLakeConnectorTest can't those types nicely. Let me handle in follow-up with `TestSetup` refactoring.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 2]
```diff
@@ -7904,8 +7904,12 @@ private List<TypeCoercionTestSetup> typeCoercionOnCreateTableAsSelectData()
                 .add(new TypeCoercionTestSetup("CHAR ' A'", "varchar", "' A'"))
                 .add(new TypeCoercionTestSetup("CHAR 'ABc'", "varchar", "'ABc'"))
                 .add(new TypeCoercionTestSetup("ARRAY[CHAR 'A']", "array(varchar)", "ARRAY['A']"))
+                .add(new TypeCoercionTestSetup("ARRAY[ARRAY[CHAR 'nested']]", "array(array(varchar))", "ARRAY[ARRAY['nested']]"))
                 .add(new TypeCoercionTestSetup("MAP(ARRAY[CHAR 'key'], ARRAY[CHAR 'value'])", "map(varchar, varchar)", "MAP(ARRAY['key'], ARRAY['value'])"))
+                .add(new TypeCoercionTestSetup("MAP(ARRAY[CHAR 'key'], ARRAY[ARRAY[CHAR 'value']])", "map(varchar, array(varchar))", "MAP(ARRAY['key'], ARRAY[ARRAY['value']])"))
+                // TODO Add test case for MAP type with ARRAY keys once https://github.com/trinodb/trino/issues/1146 is resolved
                 .add(new TypeCoercionTestSetup("CAST(ROW('a') AS ROW(x CHAR))", "row(x varchar)", "CAST(ROW('a') AS ROW(x VARCHAR))"))
+                .add(new TypeCoercionTestSetup("CAST(ROW(ROW('a')) AS ROW(x ROW(y CHAR)))", "row(x row(y varchar))", "CAST(ROW(ROW('a')) AS ROW(x ROW(y VARCHAR)))"))
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:7912]: https://github.com/trinodb/trino/issues/22042

**Pull request merged**